chainer/dataset/convert.py

```diff
@@ -139,3 +145,150 @@ def _concat_arrays_with_padding(arrays, padding):
+            indicates the host memory (CPU). If it is omitted, the array is
+            left in the original device. Asynchronous data transfer is used
+            only when device ID >= 0.
+        stream (cupy.cuda.Stream): CUDA stream. An array is sent to GPU
+            asynchronously using this stream. If ``None``, asynchronous data
+            transfer is not used.
+    """
+
+    def __init__(self, device=None, stream=None):
+        self.device = device
+        self.stream = stream
+        self.array_set = [[None, None], [None, None]]
+
+    def __call__(self, array):
+        if self.device is None or self.device < 0 or self.stream is None:
+            return to_device(self.device, array)
+
+        pin_array, cp_array = self.array_set.pop(0)
+        if pin_array is not None:
+            # Check if the size of the array matches the size of previously
+            # created and retained array.
+            if pin_array.nbytes != array.nbytes:
+                pin_array = None
+
+        with cuda.get_device_from_id(self.device):
+            if pin_array is None:
+                # Allocate memory for numpy arrays with pinned memory
+                # and cupy arrays. These arrays are retaind at self.array_set
+                # and will be reused for subsequent batches as long as
+                # the size are the same.
+
+                # The global synchronization below is necceary to ensure ALL
+                # operations including compute and data transfer submitted
+                # to GPU so far have been completed, in order to avoid possible
+                # memory corruption due to race condition among operations that
+                # use different CUDA streams.
+                # You can also solve this sort of race condition by preparing a
+                # memory pool for each CUDA stream and using it carefully.
+                runtime.deviceSynchronize()
+                pin_mem = cupy.cuda.alloc_pinned_memory(array.nbytes)
+                pin_array = numpy.frombuffer(pin_mem,
+                                             array.dtype,
+                                             array.size
+                                             ).reshape(array.shape)
+                cp_array = cupy.empty_like(array)
+
+            pin_array[...] = array  # copy(CPU): paged -> pinned
+            cp_array.set(pin_array, self.stream)  # copy: CPU to GPU
+
+        self.array_set.append([pin_array, cp_array])
+        return cp_array
+
+    def synchronize(self):
+        # Wait for completion of the data transfer submitted above.
+        # Global synchronizaton is used here for safer reason.
+        # If a caller function is correctly handling the synchronization,
+        # local synchronization (self.stream.synchronize()) may be enough.
+        if (self.device is not None and self.device >= 0 and
+                self.stream is not None):
+            runtime.deviceSynchronize()
```

Comment:
Is it safe to assume that the current device is same as `self.device`? I think it is better to make this method automatically switch the device, or write explicitly that this method assumes that the device is correctly switched by the caller, and fix `ConcatWithAsyncTransfer` to correctly switch the device (the latter might have lower overhead because only two `cudaSetDevice` is needed, i.e., `with get_device_from_id(self.device):`).

Reply:
I agree with you that it's safer to use `with get_device_from_id(...)`. changed code in that way.